src/ComposerConstraint.php

```diff
@@ -2,89 +2,148 @@
+   * concatenated with the string 'x-dev', and their operators are removed.
+   *
    * @return string
-   *   E.g. '2.x-dev || 3.x-dev'.
+   *   Lightning dev version. For example, if the constraint is
+   *   '^1.3.0 || ^2.3.0', it will be '1.x-dev || 2.x-dev'.
    */
   public function getLightningDev() {
     return $this->getDev([static::class, 'lightningRangeToDev']);
   }
 
   /**
+   * Returns the dev version of the constraint.
+   *
+   * In the dev version the constraint's ranges are replaced with the
+   * result of the callback.
+   *
    * @param callable $callback
+   *   Transformation to apply to each range.
    *
    * @return string
+   *   Dev version.
    */
-  private function getDev($callback) {
-    $dev = $this->constraint;
+  private function getDev(callable $callback) {
     $ranges = $this->getRanges();
+    $replace_pairs = [];
 
     foreach ($ranges as $oldRange) {
       $newRange = $callback($oldRange);
-      $dev = str_replace($oldRange, $newRange, $dev);
+      $replace_pairs[$oldRange] = $newRange;
     }
 
-    return $dev;
+    return strtr($this->constraint, $replace_pairs);
   }
 
   /**
+   * Returns the core dev version of a given range.
+   *
+   * In the core dev version the last digits of the range are replaced by
+   * the string 'x-dev', and the operators are removed.
+   *
    * @param string $range
-   *   E.g. '8.5.3'.
+   *   Range. E.g. '^8.5.3'.
+   *
    * @return string
-   *   E.g. '8.5.x-dev'.
+   *   Core dev version of range. E.g. '8.5.x-dev'.
    */
-  public static function coreRangeToDev($range) {
-    $numeric = preg_replace('/[^0-9\.]+/', NULL, $range);
-    $dev = preg_replace('/\.[0-9]+$/', '.x-dev', $numeric);
+  private static function coreRangeToDev($range) {
+    $stripped = static::stripOperators($range);
+    $dev = preg_replace('/\.[0-9]+$/', '.x-dev', $stripped);
 
     return $dev;
   }
 
   /**
+   * Returns the lightning dev version of a given range.
+   *
+   * In the lightning dev version the first digits of the range are
+   * concatenated with the string 'x-dev', and the operators are removed.
+   *
    * @param string $range
-   *   E.g. '1.3.0'.
+   *   Range. E.g. '^1.3.0'.
+   *
    * @return string
-   *   E.g. '1.x-dev'.
+   *   Lightning dev version of range. E.g. '1.x-dev'.
    */
-  public static function lightningRangeToDev($range) {
-    $numeric = preg_replace('/[^0-9\.]+/', NULL, $range);
-    $dev = preg_replace('/^([0-9])+\..*/', '$1.x-dev', $numeric);
+  private static function lightningRangeToDev($range) {
+    $stripped = static::stripOperators($range);
+    $dev = preg_replace('/^([0-9])+\..*/', '$1.x-dev', $stripped);
 
     return $dev;
   }
 
+  /**
+   * Returns the operator free version of a given range.
+   *
+   * @param string $range
+   *   Range. E.g. '^1.3.0'.
+   *
+   * @return string
+   *   Operator free version of range. E.g. '1.3.0'.
+   */
+  private static function stripOperators($range) {
```

Comment:
Do these methods need to be static? We can still call them in array callbacks if they're not static; for example, `$this->getDev([$this, 'coreRangeToDev'], $values)`